### PR TITLE
DM-43898: Create PlotInfo analysis tools plot element

### PR DESCRIPTION
**Pull request by jrmullaney**
Added a PlotInfoElement for use in element-wise analysis tools plot creation. Also made changes to GridPlot to accommodate the new PlotInfoElement.

### CODE REVIEW

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:140]: Ok. It's used to calculate a scale-independent spacing between the plotInfo title and the remaining text. I can see if there's another way to do it. What's the justification for not wanting specific elements (as opposed to all) being passed a fig object?

Comment (natelust) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:140]: I think it is a violation of separation of concerns and encapsulation. I think it will also encourage others to look at this, also pass the figure object, and have the elements make changes to the figure. If all you need in the dpi spacing or something like that either pass that, or possibly look to see if that is context we think other elements may need to know.

Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:140]: Thanks - makes total sense. I'll pass the fig scale transform instead, which is all that it's needed for.

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:60]: This is an important TODO, and something we might want to consider now. Adding logic to parse a `plotInfo` dict here duplicates logic already inside `addPlotInfo` in `plotUtils`, and introduces the possibility of a divergence down the line.

Instead of duplicating code, it seems to me that the better approach would be (as you rightly point out in this TODO) to separate out the dict-parsing logic inside `addPlotInfo` and drop it into its own (recently freed-up name) `parsePlotInfo`. That should allow it to be imported both here an inside the remainder of `addPlotInfo` as appropriate. I don't think such an exercise is beyond the scope of this ticket - what do you think?

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:94]: I realize this is copied over from the existing code, so I'm not expecting you to address it here necessarily. However, what I would do were it me is to add an optional `logger` argument into this function which is a logging instance. Then, inside this function where appropriate, add `if logger` blocks which then call `logger.info` or `logger.warning` etc when one of these triggers is `True`.

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:133]: Can be replaced with a call to the imported function once/if the above code is moved back into a new `parsePlotInfo` function inside `plotUtils`.

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:146]: We don't necessarily need to pass data/anything out, do we? If we do need to pass something out, perhaps the parsed plot info would be a good data point to have?

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:76]: "Width ratios."
"Height ratios."

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:88]: Other than moving the sentence to a new line, did something else change here?

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:116]: We probably need a couple of new validations down in `def validate` that verify that the length of supplied `height_ratios`/`width_ratios` (if given) matches the `numRows`/`numCols` config values.

Comment (leeskelvin) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:116]: What happens now if you try to construct your test plot with a plot Info, but pass too many `height_ratios`? Does it fail it a meaningful error?

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (leeskelvin) [reply python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:146]: But if nothing, this could just be a method which edits in-place.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:127]: What's the purpose of this? Is it to reserve some small fraction of space at the top of a plot if a plot info is _not_ given?

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:173]: No, I think not! Well spotted...!

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/gridPlot.py:141]: Something about this doesn't feel quite right, i.e., a special carve-out for a `PlotInfoElement` (here and below). Perhaps the `data` for a `PlotInfoElement` is the `figureInfo`, which would then bring it into line with how other elements have been designed - i.e., "data input needed to make the element work".

Comment (leeskelvin) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:141]: This may require us to re-think how we generate gridspecs somewhat though, i.e., we probably want the user to be able to specify the entire top-row in a gridplot should be reserved for use by a plot info element, rather than hard-coding it here.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (leeskelvin) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:127]: If so, please add a brief in-line comment above with your rationale. However, pending a discussion below of a potential re-write of how we handle plot info elements, this may end up disappearing in the rewrite.

**Review by leeskelvin — COMMENTED** [review]
Some initial thoughts which I hope are helpful. I agree that moving some of the plot info dict parsing login into a new `parsePlotInfo` function inside `plotUtils` probably makes a ton of sense. I see now on the Jira ticket that that was your recommendation also, so I wholeheartedly support that.

Some further comments here on whether or not we want to be handling plot info elements in a special way, which goes against the ethos of the "plot elements" system somewhat. It may be a necessary evil here however, considering that plot info is a somewhat special case.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:88]: No, I believe this was just a black edit.

Comment (leeskelvin) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:88]: Hm, interesting, black doesn't change this for me. At least, not in VSCode with this in my `settings.json`:
```
    "black-formatter.args": [
        "--line-length",
        "110",
    ],
```

File: python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/elements/plotInfoElement.py:55]: Now moved to `plotUtils.py` with updated docstrings.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:88]: This is fixed now - keeping the longer lines. Don't know what happened there!

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:116]: Validation added.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:127]: For now, this has been removed in lieue of developing a non-carve-out method of adding plotInfo if/when not explicitly specified.

File: python/lsst/analysis/tools/actions/plot/gridPlot.py
Comment (jrmullaney) [reply python/lsst/analysis/tools/actions/plot/gridPlot.py:141]: This "auto-add plotInfo" feature has been removed (see "in lieue" above).

File: python/lsst/analysis/tools/actions/plot/plotUtils.py
Comment (leeskelvin) [inline python/lsst/analysis/tools/actions/plot/plotUtils.py:305]: Much cleaner!

**Comment by leeskelvin** [comment]
Thanks @jrmullaney, I think this is looking great. This ticket now focuses solely on adding the new `PlotInfoElement`, and modifying `GridPlot` to be able to work with non-equal-area grid elements, such as a `PlotInfoElement`.

As we discussed, more work probably needs to be done to move `GridSpec` onto use of `subplot_mosaic`, to allow the user complete control over placement of plot elements. Did you set up a follow-up ticket for that? If so, please do add a `# TODO: See DM.....` in-line comment on this ticket to indicate that future effort will take place inside `GridSpec`.

Were it up to me, I'd approve this PR at this time. However, I see you've asked for @natelust and @sr525 for a review also, so I'll leave this open for now for them to leave their own comments if they wish. Thanks again for putting this together; it'll be super useful!

**Comment by jrmullaney** [comment]
Thanks, @leeskelvin . I've created that new ticket (DM-44283) and added a TODO referring to it. I've also removed `atools/gridPlotTest.py` from the PR as it shouldn't be included in a merge. I'll run Jenkins now and wait for @natelust and @sr525 to sign-off or comment.
Thanks for your help on this!

**Review by leeskelvin — APPROVED** [review]

**Pull request merged**
